src/storage/src/storage.rs

```diff
@@ -49,71 +49,114 @@ use crate::{Error, Result};
+    /// more bytes as persisted than in the latest report. This could indicate:
+    /// - a corrupted message from the service, either the earlier message
+    ///   reporting more bytes persisted than actually are, or the current
+    ///   message indicating fewer bytes persisted.
+    /// - a bug in the service, where it reported bytes as persisted when they
+    ///   were not.
+    /// - a bug in the client, maybe storing the incorrect byte count, or
+    ///   parsing the messages incorrectly.
+    ///
+    /// All of these conditions indicate a bug, and in Rust it is idiomatic to
+    /// `panic!()` when a bug is detected. However, in this case it seems more
+    /// appropriate to report the problem, as the client library cannot
+    /// determine the location of the bug.
     #[error(
         "the service previously persisted {offset} bytes, but now reports only {persisted} as persisted"
     )]
     UnexpectedRewind { offset: u64, persisted: u64 },
 
+    /// The service reports more bytes persisted than sent.
+    ///
+    /// # Troubleshoot
+    ///
+    /// Most likely this indicates that two concurrent uploads are using the
+    /// same session. Review your application design to avoid concurrent
+    /// uploads.
+    ///
+    /// It is possible that this indicates a bug in the service, client, or
+    /// messages corrupted in transit.
     #[error("the service reports {persisted} bytes as persisted, but we only sent {sent} bytes")]
     TooMuchProgress { sent: u64, persisted: u64 },
-}
 
-/// The error type for checksum comparisons.
-///
-/// By default the client library computes a checksum of the uploaded data, and
-/// compares this checksum against the value returned by the service.
-///
-/// # Example
-/// ```
-/// # use google_cloud_storage::client::Storage;
-/// # use google_cloud_storage::ChecksumMismatch;
-/// # async fn sample(client: &Storage) -> anyhow::Result<()> {
-/// use std::error::Error as _;
-/// let upload = client
-///     .upload_object("projects/_/buckets/my-bucket", "my-object", "hello world")
-///     .with_if_generation_not_match(0);
-/// match upload.send().await {
-///     Ok(object) => println!("Successfully uploaded the object"),
-///     Err(error) if error.is_serialization() => {
-///         println!("Some problem {error:?} sending the data to the service");
-///         if let Some(m) = error.source().and_then(|e| e.downcast_ref::<ChecksumMismatch>()) {
-///             println!("The checksums did not match: {m}");
-///         }
-///     },
-///     Err(e) => return Err(e.into()), // not handled in this example
-/// }
-/// # Ok(()) }
-/// ```
-///  
-/// # Troubleshooting
-///
-/// Data integrity problems are notoriously difficult to root cause. If you are
-/// using pre-existing, or pre-computed checksum values, you may want to verify
-/// the source data.
-#[derive(thiserror::Error, Debug)]
-#[non_exhaustive]
-pub enum ChecksumMismatch {
```

Comment:
#2792 wants to use this.

Reply:
Refactored the code to have a common public enum for uploads and downloads.